### DISCUSSION TITLE
Grid line painter difference between Dark and Light themes when spanning

### DISCUSSION THREAD

**Discussion by jmenzies12** [eclipse-nattable/nattable#67] [Q&A] [answered]
Hey there,

I am in the process of trying to configure our NatTables to display nicely with the dark theme. I ran into a weird issue however, even with the default painter, it seems to be displaying lines for all of the rows, even ones that should be hidden due to spanning. However it is not doing this on light theme (screenshots below). I tried using a custom painter and setting them to blue to see if maybe it was just not showing up on light mode but it has the same issue.

It wouldn't surprise me at all if we've configured something wrong but I'm not entirely sure what would cause this.

Any thoughts?

![image](https://github.com/eclipse-nattable/nattable/assets/3521863/a4183fc3-3fa6-40d6-9c45-d8cd328b1e50)

![image](https://github.com/eclipse-nattable/nattable/assets/3521863/cb040aaf-96ba-464d-b0fc-c11ab0641fde)

Cheers
Jonathan

**Comment by fipro78** [comment]
I suppose you have created your own dark theme and you are not using the `DarkNatTableThemeConfiguration`?

I would expect that you configured a body cell painter without background painting in your dark theme. The grid lines are actually always painted. On spanning the grid lines are "overpainted" by the cell background, either via `BackgroundPainter` or by `TextPainter#paintBg`. Without painting the background in the body cells, the grid lines will "shine through".

**Reply by jmenzies12** [reply]
Thanks for the response! I had a look into that, after removing everything I could find so I could add things back in, it seems that it's coming from the AbstractCSSEngine, I installed the NatTable extension.e4 feature so that it would work with Eclipse's themes.

When I change the eclipse theme to dark, it seems to be registering a TextPainter (with the background painting off) with the config label "source", this is the first label in the configLabels list, so when the SpanningDataLayer goes to retrieve the painter it gets that one first.

Wondering if there is a way to configure this?

**Reply by fipro78** [reply]
What do you want to configure? Do you have a a CSS configuration for NatTable in your project? Then configure it there.

AFAIK there is no default CSS configuration for NatTable anywhere. At least I am not aware of it. And we also do not ship a default CSS configuration. We only have the one in our example application:

https://github.com/eclipse-nattable/nattable/blob/master/org.eclipse.nebula.widgets.nattable.examples.e4%2Fcss%2Fdefault.css

**Reply by jmenzies12** [reply]
Sorry for the confusion. I will clarify what I'm seeing below.

We are using the org.eclipse.nebula.widgets.nattable.extension.e4 feature that hooks up NatTable with the Eclipse theme settings for our Eclipse RCP application.

When I include this feature and then turn on dark mode. What happens is that the AbstractCSSEngine in Eclipse triggers the NatTableCSSHandler, in this class the method NatTableCSSHandler.onAllCSSPropertiesApplyed then registers a TextPainter with paintBG set to false in the config registry for the natTable with the configLabel set to "source". Because of this, this painter overrides any painter that I register and does not paint the background in order to overpaint the grid lines.

Do I have to register a CSS configuration to override this behaviour?

**Reply by fipro78** [reply, marked as answer]
> Because of this, this painter overrides any painter that I register and does not paint the background in order to overpaint the grid lines.

Yes, the CSS engine always wins over any style configuration. This is the nature of the CSS engine to support dynamic changes to CSS.

> Do I have to register a CSS configuration to override this behaviour?

That is what I was trying to tell you. There is no default CSS configuration for a NatTable. IIRC it will return some default if no configuration is found. And in your case, this is a simple TextPainter without a background. Why this is the case for your Eclipse themes? I have no idea. Maybe because you configured some values in the light mode and missed them in the dark mode? It will always use a TextPainter with paintBG false, because it will interprete a background CSS configuration to add a BackgroundPainter. If no background value is found, no BackgroundPainter is added.

I don't know if you have a CSS configuration for the NatTable in your dark theme or not. But the `CellPainterFactory` is used to retrieve a painter. There you can see how a painter stack is created based on which css properties.

The CSS capabilities are described in the New & Noteworthy of 1.4:  
https://eclipse.dev/nattable/nandn/nandn_140.php
